portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,70 +27,107 @@
+        // Add fun facts first to factsAndComments return variable.
+        List<String> factsAndComments = new ArrayList<>();
+        factsAndComments.addAll(funFacts);        
         
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         Query query = new Query("siteComments").addSort("timeStamp", SortDirection.DESCENDING);
         PreparedQuery results = datastore.prepare(query);
-
+ 
         List<String> pastComments = new ArrayList<>();
+        int validNumRequest = 0;
+        int numCommentsWanted = 0;
+        boolean numCommentsWantedTrigger = true;
+ 
         for (Entity entity : results.asIterable()) {
-            String comment = (String) entity.getProperty("newComment");
-            pastComments.add(comment);
+            validNumRequest = Integer.parseInt((String)(entity.getProperty("seeNumComments")));
+            if (validNumRequest >= 0){
+                String comment = (String) (entity.getProperty("newComment"));
+
+                //For testing queried result comments
+                // System.out.println("Query result comment: " + comment);
+
+                pastComments.add(comment);
+            } else {
+                System.out.println("Input of number comments wanted not greater than or equal to 0: " + validNumRequest);
+            }
+           
+            //Grab number of comments wanted to be seen by user only in the first entity
+            while (numCommentsWantedTrigger) {
+                numCommentsWanted = validNumRequest;
+                numCommentsWantedTrigger = false;
+            }
+        }
+ 
+        if (numCommentsWanted > pastComments.size()) {
+            numCommentsWanted = pastComments.size();           
         }
+ 
+        pastComments = pastComments.subList(0, numCommentsWanted);
 
-        List<String> factsAndComments =  new ArrayList<>();
-        factsAndComments.addAll(funFacts);
-        factsAndComments.addAll(pastComments);
+        // For testing proper number and right comments are in sublist
+        // for (String comment : pastComments) {
+        //     System.out.println("sublist comment: " + comment);
+        // }
 
+        factsAndComments.addAll(pastComments);
+ 
         response.setContentType("application/json;");
         Gson gson = new Gson();
         response.getWriter().println(gson.toJson(factsAndComments));
     }
-
+ 
     /** Process POST request form input, add to pastComments, and redirect to index.html */
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        // Get the input from the form.
-        String comment = getParameter(request, "text-input", "");
         long timestamp = System.currentTimeMillis();
-
+        
+        // Get the input from the form.
+        String comment = getParameter(request, "text-input", ""); 
+        String numCommentsString = getParameter(request, "num-comments", "0");
```

Comment:
You're eventually going to try to parse this as a number, right?  So perhaps it'd be a good idea to validate that the input is good here, so someone doesn't get bad data into your datastore, which will begin crashing the page.

Looking ahead, it seems like your form is already specifying that the input should be a number, so at least someone couldn't do this accidentally.  It's still good practice, though, to validate your input on the server, so someone can't send you a manual post request with something malicious.

Reply:
That makes me wonder, I was thinking I should store all entered input to track website interaction including invalid input. Let me know what is best as you said above that bad data can eventually crash the page.

Reply:
I think it's generally better practice to not store invalid input.  Doing a "proper" job of it would involve validating the form as the user is typing into it, and not allowing them to submit if they've got something wrong.  Then, the server can just reject the request if it's got invalid input, because it must have come from someone trying to attack the service, rather than user error.

I'm not too worried about having you validate the form in js, but rejecting bad input on the server seems like a good idea to implement, especially if that bad data would cause your page to stop loading.